crates/shared/src/code_simulation.rs

```diff
@@ -0,0 +1,190 @@
+//! Abstraction for simulating calls with overrides.
+
+use crate::{
+    tenderly_api::{SimulationKind, SimulationRequest, TenderlyApi},
+    transport::extensions::{EthExt as _, StateOverrides},
+    Web3,
+};
+use anyhow::{bail, Context as _, Result};
+use std::sync::Arc;
+use web3::types::{BlockNumber, CallRequest};
+
+/// Simulate a call with state overrides.
+#[async_trait::async_trait]
+pub trait CodeSimulating: Send + Sync + 'static {
+    async fn simulate(&self, call: CallRequest, overrides: StateOverrides) -> Result<Vec<u8>>;
```

Comment:
Does it make sense to have the result type differentiate "node error" from "evm error"?

Reply:
Yes, this is a good idea. I think, we can optimistically allow price estimages if there is a node/transport/http error.

Reply:
Created https://github.com/cowprotocol/services/issues/528